usdShade: Include connectable validation from usdchecker PrimEncapsulationRule

beersandrew opened the pull request:
### Description of Change(s)
- adding necessary validation files for connectable validation (no nested shaders, no shaders inside scopes)

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3?pane=issue&itemId=65424985

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: add connectable validator

- added plugInfo, tokens, validation & test related to connectable nesting rules in usdShade

jesschimein commented:
Filed as internal issue #USD-9979

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -30,6 +30,70 @@
+
+    const UsdShadeConnectableAPI& connectable = UsdShadeConnectableAPI(usdPrim);
+
+    if (!usdPrim.GetTypeName().IsEmpty() && connectable) {
```
nit: you can negate this and do an early return. And then you can move this entire logic outside the if block.

tallytalwar replied:
Also no need to do check for primType here as well.

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -30,6 +30,70 @@
+
+        if (parentPrim && !parentPrim.IsPseudoRoot()) {
+            UsdShadeConnectableAPI parentConnectable = UsdShadeConnectableAPI(parentPrim);
+            if (parentPrim.GetTypeName().IsEmpty()){
```
How about you use ternary operator here:

```
UsdShadeConnectableAPI parentConnectable = (parentPrim.GetTypeName().IsEmpty()) ? UsdShadeConnectableAPI() : UsdShadeConnectableAPI(parentPrim)
```

But why do you want to have this check to begin with? parentConnectable for prim with no typename should return false for IsContainer, so I think you can simply remove this parentPrim typename is empty check.
```

tallytalwar replied:
I see you are doing this because of the comment in complianceChecker to work around a bug! I think that comment is outdated, the bug its talking about is all fixed now.

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -30,6 +30,70 @@
+                parentConnectable = UsdShadeConnectableAPI();
+            }
+
+            if (parentConnectable && !parentConnectable.IsContainer()) {
```
I will retain the original code comments. Code comments are great as it makes sure the intent being described in the code comment matches in the code. Specially for a reviewer or a future developer to work with.

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -30,6 +30,70 @@
+            }
+            else if (!parentConnectable) {
+                UsdPrim currentAncestor = parentPrim.GetParent();
+                while(currentAncestor && !currentAncestor.IsPseudoRoot()) {
```
You can maybe break this while loop to a recursive lambda?

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -30,6 +30,70 @@
+                                            UsdValidationErrorSite(usdPrim.GetStage(),
+                                                                   usdPrim.GetPath())
+                                    },
+                                    TfStringPrintf("Connectable %s <%s> can only have"
```
Will be good to retain the code comment from UsdUtilsComplianceChecker here.

tallytalwar commented on pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp:
```diff
@@ -410,6 +412,55 @@ TestUsdShadeMaterialBindingAPIAppliedValidator()
+    const UsdShadeShader& insideShader = UsdShadeShader::Define(usdStage, SdfPath("/RootMaterial/Shader/InsideShader"));
+
+    // Verify error that does not allow a connectable to be parented by non-container connectable
+    UsdValidationErrorVector errors = validator->Validate(insideShader.GetPrim());
```
Might also make sense to have a Validate for "/RootMaterial/Shader" and show no errors are reported for the same.

tallytalwar reviewed (commented):
Thanks @beersandrew . Left a couple of notes to be addressed.

beersandrew committed:
refactor: various fixes based on PR comments

- added a successful test to verify 0 errors
- reduced nesting in validation code
- added comments in from the complianceChecker code
- removed checking TypeName as the code that required that check has been fixed
- updated ancestor check code to be more clear

beersandrew commented:
Thanks for the notes here @tallytalwar, I think I've addressed them all
